core/src/main/java/com/datastax/oss/driver/internal/core/util/collection/QueryPlan.java

```diff
@@ -0,0 +1,108 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.util.collection;
+
+import com.datastax.oss.driver.api.core.loadbalancing.LoadBalancingPolicy;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.session.Request;
+import com.datastax.oss.driver.api.core.session.Session;
+import com.datastax.oss.driver.internal.core.loadbalancing.DefaultLoadBalancingPolicy;
+import com.datastax.oss.driver.shaded.guava.common.collect.Iterators;
+import java.util.AbstractCollection;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicInteger;
+import net.jcip.annotations.ThreadSafe;
+
+/**
+ * A specialized, thread-safe queue implementation for {@link
+ * LoadBalancingPolicy#newQueryPlan(Request, Session)}.
+ *
+ * <p>All nodes must be provided at construction time. After that, the only valid mutation operation
+ * is {@link #poll()}, other methods throw.
+ *
+ * <p>This class is not a general-purpose implementation, it is tailored for a specific use case in
+ * the driver. It makes a few unconventional API choices for the sake of performance (see {@link
+ * #QueryPlan(Object...)}. It can be reused for custom load balancing policies; if you plan to do
+ * so, study the source code of {@link DefaultLoadBalancingPolicy}.
+ */
+@ThreadSafe
+public class QueryPlan extends AbstractCollection<Node> implements Queue<Node> {
+
+  private final Object[] nodes;
+  private final AtomicInteger nextIndex = new AtomicInteger();
+
+  /**
+   * @param nodes the nodes to initially fill the queue with. For efficiency, there is no defensive
+   *     copy, the provided array is used directly. The declared type is {@code Object[]} because of
+   *     implementation details of {@link DefaultLoadBalancingPolicy}, but all elements must be
+   *     instances of {@link Node}, otherwise instance methods will fail later.
+   */
+  public QueryPlan(Object... nodes) {
+    this.nodes = nodes;
+  }
+
+  @Override
+  public Node poll() {
+    // We don't handle overflow. In practice it won't be an issue, since the driver stops polling
+    // once the query plan is empty.
+    int i = nextIndex.getAndIncrement();
+    return (i >= nodes.length) ? null : (Node) nodes[i];
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>The returned iterator reflects the state of the queue at the time of the call, and is not
+   * affected by further modifications.
+   */
+  @Override
+  public Iterator<Node> iterator() {
```

Comment:
From what I can tell, we never use `iterator()` on `QueryPlan` coming back from `newQueryPlan`, so this implementation would ultimately not be used and is more for completeness of implementation right?

Reply:
I didn't check but it might be used in tests. Also, `AbstractCollection.toString` relies on it, and you need `toString` for debugging or logs.